File: project_hft/rsi/rsi_parallel.c
// rsi_parallel.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <omp.h>

#define PERIOD 14
#define N 1000000

float* generate_random_prices(int n) {
    float* prices = malloc(sizeof(float) * n);
    if (!prices) return NULL;

    prices[0] = 100.0f;
    for (int i = 1; i < n; ++i) {
        float change = ((rand() % 2001) - 1000) / 1000.0f;
        prices[i] = prices[i - 1] + change;
    }
    return prices;
}

void calc_rsi_parallel(float* prices, float* rsi, int n, int period) {
    float gain = 0.0f, loss = 0.0f;

    for (int i = 1; i <= period; ++i) {
        float change = prices[i] - prices[i - 1];
        if (change > 0) gain += change;
        else loss -= change;
    }

    float avg_gain = gain / period;
    float avg_loss = loss / period;

    #pragma omp parallel for
    for (int i = period + 1; i < n; ++i) {
        float change = prices[i] - prices[i - 1];
        float g = change > 0 ? change : 0;
        float l = change < 0 ? -change : 0;

        float local_avg_gain = (avg_gain * (period - 1) + g) / period;
        float local_avg_loss = (avg_loss * (period - 1) + l) / period;

        float rs = local_avg_loss == 0 ? 100.0f : local_avg_gain / local_avg_loss;
        rsi[i] = 100.0f - (100.0f / (1.0f + rs));
    }
}

int main() {
    srand(time(NULL));
    float* prices = generate_random_prices(N);
    float* rsi = calloc(N, sizeof(float));

    double start = omp_get_wtime();
    calc_rsi_parallel(prices, rsi, N, PERIOD);
    double end = omp_get_wtime();

    printf("Parallel Time: %.6f sec\n", end - start);
    printf("Sample RSI: %.2f %.2f %.2f\n", rsi[N/2], rsi[N/2 + 1], rsi[N/2 + 2]);

    free(prices);
    free(rsi);
    return 0;
}

File: project_hft/rsi/rsi_parallel.py
from multiprocessing import Process, Queue
import time
import random

# ---------- Stage 1: Price Difference ----------
def stage_diff(input_q, output_q):
    prev = None
    index = 0
    while True:
        item = input_q.get()
        if item == "DONE":
            output_q.put("DONE")
            break
        if prev is not None:
            diff = item - prev
            output_q.put((index, diff))
            index += 1
        prev = item

# ---------- Stage 2: Gain and Loss ----------
def stage_gain_loss(input_q, output_q):
    while True:
        item = input_q.get()
        if item == "DONE":
            output_q.put("DONE")
            break
        i, diff = item
        gain = max(diff, 0)
        loss = max(-diff, 0)
        output_q.put((i, gain, loss))

# ---------- Stage 3: EMA of Gain and Loss ----------
def stage_avg(input_q, output_q, period):
    gain_hist = []
    loss_hist = []
    EMA_gain = None
    EMA_loss = None
    while True:
        item = input_q.get()
        if item == "DONE":
            output_q.put("DONE")
            break
        i, gain, loss = item
        if i < period:
            gain_hist.append(gain)
            loss_hist.append(loss)
            continue
        elif i == period:
            EMA_gain = sum(gain_hist + [gain]) / period
            EMA_loss = sum(loss_hist + [loss]) / period
        else:
            EMA_gain = (EMA_gain * (period - 1) + gain) / period
            EMA_loss = (EMA_loss * (period - 1) + loss) / period

        output_q.put((i, EMA_gain, EMA_loss))

# ---------- Stage 4: RSI Calculation ----------
def stage_rsi(input_q, output_q):
    while True:
        item = input_q.get()
        if item == "DONE":
            output_q.put("DONE")
            break
        i, ag, al = item
        rs = ag / al if al != 0 else float('inf')
        rsi = 100 - (100 / (1 + rs))
        output_q.put((i, rsi))

# ---------- Pipeline Driver ----------
def rsi_pipeline(close_prices, period=14):
    q1 = Queue()
    q2 = Queue()
    q3 = Queue()
    q4 = Queue()
    q_out = Queue()

    # Spawn stage processes
    p1 = Process(target=stage_diff, args=(q1, q2))
    p2 = Process(target=stage_gain_loss, args=(q2, q3))
    p3 = Process(target=stage_avg, args=(q3, q4, period))
    p4 = Process(target=stage_rsi, args=(q4, q_out))

    for p in [p1, p2, p3, p4]:
        p.start()

    # Feed input prices
    for cp in close_prices:
        q1.put(cp)
    q1.put("DONE")

    # Collect output RSI values
    result = [None] * len(close_prices)
    while True:
        item = q_out.get()
        if item == "DONE":
            break
        i, rsi = item
        result[i] = rsi

    for p in [p1, p2, p3, p4]:
        p.join()

    return result

# ---------- Main ----------
if __name__ == "__main__":
    # Large price dataset for stress test
    large_prices = [100 + random.uniform(-1, 1) for _ in range(1000000)]

    start_time = time.time()
    rsi_result = rsi_pipeline(large_prices, period=14)
    end_time = time.time()

    print(f" Parallel RSI (Pipelined) Execution Time: {end_time - start_time:.4f} seconds")
    # Optional: print first 10 RSI values (excluding None values)
    print("Sample RSI values:", [r for r in rsi_result if r is not None][:10])
